### PR TITLE
ECG DATA QUALITY pull request

### PR DESCRIPTION
**Pull request by aungkonazim**
Hi Dr. Hnat,

I have done the ECG data quality computation within the wrapper class.
Since I don't have the  access to data I couldn't run the computation. Although it seems good but without running I can't be too confident on my code.

Can you please run the code at your convenience and let me know. I hope to make everything written by today including respiration code. To allow me make some changes to the constant definitions as Dr. Kumar mentioned I urge you to give me some time in tomorrow morning.

Thanks and Regards
Azim

### CODE REVIEW

File: src/main/java/md2k/mcerebrum/cstress/autosense/AUTOSENSE.java [hunk 1]
```diff
@@ -49,6 +49,13 @@
     public static final int QUALITY_MISSING = 4;
     public static final int QUALITY_BAND_LOOSE = 2;
 
+    public static final int QUALITY_acceptableOutlierPercent = 34;
+    public static final int QUALITY_outlierThresholdHigh = 4000;
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/autosense/AUTOSENSE.java]: Most of these should be based on a ratio (e.g. 0 <= x <= 1) and combined with the ADC range in the logic.  This ADC should be specified as an input to the processing module

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 2]
```diff
@@ -50,8 +69,22 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
         DataPointStream ecg = datastreams.getDataPointStream(StreamConstants.ORG_MD2K_CSTRESS_DATA_ECG);
         DataPointStream ecgQuality = datastreams.getDataPointStream(StreamConstants.ORG_MD2K_CSTRESS_DATA_ECG_QUALITY);
 
-        ECGQualityCalculation ecgComputation = new ECGQualityCalculation(3, 50, 4500, 20, 2, 47);
-        List<DataPoint> quality = ecgComputation.computeQuality(ecg.data, 5000); //0.67
+        this.acceptableOutlierPercent = AUTOSENSE.QUALITY_acceptableOutlierPercent;
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Refactor to just call the constants where needed.  There is no need to introduce additional variables to do this

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 3]
```diff
@@ -50,8 +69,22 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+        this.outlierThresholdHigh = AUTOSENSE.QUALITY_outlierThresholdHigh;
+        this.outlierThresholdLow = AUTOSENSE.QUALITY_outlierThresholdLow;
+        this.ecgThresholdBandLoose = AUTOSENSE.QUALITY_ecgThresholdBandLoose;
+        this.bufferLength = AUTOSENSE.QUALITY_bufferLength;
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: What is the purpose of a constant 2000 unit buffer?  If this is arbitrary, please use something like ArrayList to handle this without forcing.  See line 87 for an example.

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 4]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+                result.add(new DataPoint(dpA[0].timestamp, currentQuality(data)));
+            }
+        }
+
+
+        return result;
+
+    }
+    private int currentQuality(int[] data) {
+        classifyDataPoints(data);
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Methods should be written to not have side-effects.  This clearly does due to it not having a return value.

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 5]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+        classifySegment(data);
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Methods should be written to not have side-effects.  This clearly does due to it not having a return value.

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 6]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+
+        classBuff[(classHead++) % classBuff.length] = segment_class;
+        envelBuff[(envelHead++) % envelBuff.length] = max_value - min_value;
+        classifyBuffer();
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Methods should be written to not have side-effects.  This clearly does due to it not having a return value.

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 7]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+
+        if (segment_class == AUTOSENSE.SEGMENT_BAD) {
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: segment_class is not defined within this method, further indicating some sort of global state

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 8]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+            return AUTOSENSE.QUALITY_BAND_OFF;
+        } else if (2 * amplitude_small > envelBuff.length) {
+            return AUTOSENSE.QUALITY_BAND_LOOSE;
+        }else if(max_value - min_value <= ecgThresholdBandLoose) {
+            return AUTOSENSE.QUALITY_BAND_LOOSE;
+        }
+        return AUTOSENSE.QUALITY_GOOD;
+    }
+
+    private void classifyBuffer() {
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Should return something

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 9]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+        amplitude_small = 0;
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: For Java, please follow the CamelCase variable naming scheme

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 10]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+        for (int i = 0; i < envelBuff.length; i++) {
+            if (envelBuff[i] < ecgThresholdBandLoose) {
+                amplitude_small++;
+            }
+        }
+    }
+
+    private void classifySegment(int[] data) {
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Should return something.

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 11]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+        int outliers = large_stuck + large_flip + small_stuck + discontinuous ;
+        if (100 * outliers > acceptableOutlierPercent * data.length) {
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Why multiple by 100 here?

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 12]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+            segment_class = AUTOSENSE.SEGMENT_BAD;
+        } else {
+            segment_class = AUTOSENSE.SEGMENT_GOOD;
+        }
+    }
+
+    private void classifyDataPoints(int[] data){
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: This should return something

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 13]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+        small_stuck=0;
+        large_flip=0;
+        discontinuous=0;
+        max_value=data[0];
+        min_value=data[0];
+        for(int i=0;i<data.length;i++){
+            int im=((i==0)?(data.length-1):(i-1));
+            int ip=((i==data.length-1)?(0):(i+1));
+            boolean stuck=((data[i]==data[im])&&(data[i]==data[ip]));
+            boolean flip=((Math.abs(data[i]-data[im])>4000)||(Math.abs(data[i]-data[ip])>4000));
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Constant values should not be here

File: src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java [hunk 14]
```diff
@@ -70,4 +103,88 @@ public ECGDataQuality(DataStreams datastreams, double qualityThreshold) {
...
+            boolean disc=((Math.abs(data[i]-data[im])>100)||(Math.abs(data[i]-data[ip])>100));
```
Comment (twhnat) [inline src/main/java/md2k/mcerebrum/cstress/features/ECGDataQuality.java]: Constant values should not be here

**Review by twhnat — CHANGES_REQUESTED** [review]
In general, code should be written to be side-effect free.  Please update your algorithm to resolve these and other noted issues.  We can discuss in the morning if necessary.

**Comment by aungkonazim** [comment]
Hi Dr. Hnat,

Would you please look at the code at your convenience. I have finalized the code for both ECG and RIP.

Thanks,
Azim

**Review by twhnat — APPROVED** [review]

**Pull request merged**